src/v/raft/consensus.cc

```diff
@@ -1740,7 +1740,9 @@ ss::future<> consensus::hydrate_snapshot() {
 
 ss::future<> consensus::truncate_to_latest_snapshot() {
     auto lstats = _log.offsets();
-    if (lstats.start_offset > _last_snapshot_index) {
+    if (
+      lstats.start_offset > _last_snapshot_index
+      && lstats.dirty_offset == _last_snapshot_index) {
```

Comment:
Condition looks a bit brittle and hard to understand. Maybe we shouldn't do early exit here at all and let the stages do their own early exits just in case one of them didn't finish.